### ISSUE TITLE
java.lang.AssertionError: illegal type variable reference only happend on Android N or higher versions

### ISSUE THREAD

**Issue by chengdroid** [orhanobut/hawk#195]
Please try to fill all questions below before submitting an issue.

- [ ] On which android versions do you have this issue? :Android 7.0 and 7.1.1 
- [ ] On which phones do you have this issue? :SamSung S7edge, HUAWEI mate8, HUAWEI P9
- [ ] Which hawk version are you using? : hawk 1.18
- [ ] Does this issue happen always or is it flaky? : always on Android 7.0 and 7.1.1
- [ ] Can you write failing test? : 
- [ ] If it is Hawk.put and Hawk.get issue, can you write down which data you were trying to save and get it back? 
- [ ] Is the data you are trying to save it huge or small?

Because my English is not good ,so i describe it  easily. many people said it caused by Android proguard,but i did not use Android proguard with my app. this error only happend on Android versions 7.0 and 7.1.1,it runs OK on Android 6.0 or lower versions.here is log:
```

                                                   java.lang.AssertionError: illegal type variable reference
                                                       at libcore.reflect.TypeVariableImpl.resolve(TypeVariableImpl.java:111)
                                                       at libcore.reflect.TypeVariableImpl.getGenericDeclaration(TypeVariableImpl.java:125)
                                                       at libcore.reflect.TypeVariableImpl.hashCode(TypeVariableImpl.java:47)
                                                       at java.util.Arrays.hashCode(Arrays.java:4153)
                                                       at com.google.gson.internal.$Gson$Types$ParameterizedTypeImpl.hashCode($Gson$Types.java:479)
                                                       at com.google.gson.reflect.TypeToken.<init>(TypeToken.java:64)
                                                       at com.orhanobut.hawk.HawkEncoder$1.<init>(HawkEncoder.java:113)
                                                       at com.orhanobut.hawk.HawkEncoder.toList(HawkEncoder.java:113)
                                                       at com.orhanobut.hawk.HawkEncoder.decodeNew(HawkEncoder.java:95)
                                                       at com.orhanobut.hawk.HawkEncoder.decode(HawkEncoder.java:58)
                                                       at com.orhanobut.hawk.Hawk.get(Hawk.java:136)
                                                       at com.ruihai.xingka.api.model.AccountInfo.getUserTags(AccountInfo.java:159)
                                                       at com.ruihai.xingka.ui.mine.EditUserDataActivity.onCreate(EditUserDataActivity.java:139)
                                                       at android.app.Activity.performCreate(Activity.java:6910)
                                                       at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1123)
                                                       at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2746)
                                                       at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2864)
                                                       at android.app.ActivityThread.-wrap12(ActivityThread.java)
                                                       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1567)
                                                       at android.os.Handler.dispatchMessage(Handler.java:105)
                                                       at android.os.Looper.loop(Looper.java:156)
                                                       at android.app.ActivityThread.main(ActivityThread.java:6523)
                                                       at java.lang.reflect.Method.invoke(Native Method)
                                                       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:941)
                                                       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:831)
```
here is the code of AccountInfo.java:  
```
...  
public List<UserTag> getUserTags() {
        return Hawk.get(KEY_USER_TAG);
    }
...
```
here is the code of EditUserDataActivity.onCreate:
```
   @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_editdata);
...
        currentUser = AccountInfo.getInstance().loadAccount();
        List<UserTag> userTags = AccountInfo.getInstance().getUserTags();
        List<String> tagNames = new ArrayList<>();
...
```
Can any people help me .thank you very much.

**Comment by aiwiguna** [comment]
Hi, i am have this problem too
On which android versions do you have this issue? :Android 7.0 and 8.0
 On which phones do you have this issue? : Nexus 9 and 6P
 Which hawk version are you using? : hawk 1.19 and 1.23
 Does this issue happen always or is it flaky? : always
 Can you write failing test? : this problem happen when i am updating SDK and library in my project gradle
before -> no error
    compileSdkVersion 25
    buildToolsVersion "25.0.3"
    hawk 1.19
after -> error
    compileSdkVersion 26
    buildToolsVersion "26.0.2"
    hawk 1.19 or 1.23 (same error)

 If it is Hawk.put and Hawk.get issue, can you write down which data you were trying to save and get it back?
i am saving a list of my model Order
public class Order {
    long orderId;
    int count;
    String product;
    String type;
    Quantity productQty;
    int border;
    int frame;
    List<String> sourceImages;
    List<String> edittedImages;
    List<String> urlImage;
}
 Is the data you are trying to save it huge or small?

**Comment by knife-s** [comment]
is there any update on this?

**Comment by kkl260** [comment]
I'm getting this as well. @knife-s If you remove `debuggable true` from your build.gradle file you'll be good. Of course, you won't be able to debug your app. Would be nice to find a better fix for this.

**Comment by sraiteri** [comment]
@kkl260 & @knife-s,

I've been facing a similar problem with Hawk after upgrading from Android Gradle Plugin 2.3.3 to Android Gradle Plugin 3.3.0. Seems like the workaround of **debuggable true** (with proguard enabled) fixes crashes on API 24, 25 & 26.

I've been doing a bit more digging, and since Gson 2.8.0 the `TypeToken` has been enriched with a `.getParameterized` method which replaces some of the generic `TypeToken` creation happening in the guts of `HawkConverter.java`.

With the latest Gson dependency at `2.8.2` in master, there's already an open PR (https://github.com/orhanobut/hawk/pull/215) to implement this. If I take latest master branch w/ these changes, I find those changes fix this particular crash in my Android app.

I've included a custom build Hawk .aar & included Facebook Conceal `1.1.3` .aar in my case. I'm not sure if it's an issue with ProGuard being to aggressive, but there's plenty of `ClassNotFound` exceptions if I don't force it in the `build.gradle` file.

```groovy
// Hawk (for encrypted persistent storage)
    // implementation "com.orhanobut:hawk:2.0.1"
    // TODO: Including Hawk and Facebook Conceal .aar dependencies is a workaround to fix a crash on SDK 24, 25 & 26.
    // Once a new version of Hawk is published (pending 2.1.0??) this should be changed back to the Hawk dependency only.
    implementation project(":hawk-2.1.0-custom")
    implementation "com.facebook.conceal:conceal:1.1.3:@aar"
```  

**Hawk 2.1.0 + `TypeToken.getParameterized` change**
[hawk-2.1.0-custom.aar.zip](https://github.com/orhanobut/hawk/files/2776281/hawk-2.1.0-custom.aar.zip)

@orhanobut - any chance of getting this PR merged for version `2.1.0`?

**Comment by kissmiao** [comment]
@sraiteri  谢谢你，使用你的hawk-2.1.0-custom.aar.zip，解决了我的问题
